libtransmission/blocklist.h

```diff
@@ -9,19 +9,93 @@
 #error only libtransmission should #include this header.
 #endif
 
+#include <string>
+#include <vector>
+
+#include "file.h" // for tr_sys_file_t
+#include "tr-assert.h"
+#include "tr-macros.h"
+
 struct tr_address;
-struct tr_blocklistFile;
 
-tr_blocklistFile* tr_blocklistFileNew(char const* filename, bool isEnabled);
+struct BlocklistFile
+{
+public:
+    BlocklistFile(char const* filename, bool isEnabled)
+        : is_enabled_(isEnabled)
+        , filename_(filename)
+    {
+    }
+
+    ~BlocklistFile();
+
+    void close();
+
+    [[nodiscard]] bool exists() const;
+
+    [[nodiscard]] char const* getFilename() const;
+
+    // TODO: This function should be const, but cannot be const due to it calling ensureLoaded()
+    size_t getRuleCount();
+
+    [[nodiscard]] constexpr bool isEnabled() const
+    {
+        return is_enabled_;
+    }
 
-char const* tr_blocklistFileGetFilename(tr_blocklistFile const* b);
+    void setEnabled(bool isEnabled);
 
-int tr_blocklistFileGetRuleCount(tr_blocklistFile const* b);
+    bool hasAddress(tr_address const& addr);
 
-void tr_blocklistFileFree(tr_blocklistFile* b);
+    /// @brief Read the file of ranges, sort and merge, write to our own file, and reload from it
+    size_t setContent(char const* filename);
 
-void tr_blocklistFileSetEnabled(tr_blocklistFile* b, bool isEnabled);
+private:
+    struct IPv4Range
+    {
+        uint32_t begin_;
+        uint32_t end_;
+
+        /// @brief Used for std::bsearch of an IPv4 address
+        static int compareAddressToRange(void const* va, void const* vb)
+        {
+            auto const* a = reinterpret_cast<uint32_t const*>(va);
+            auto const* b = reinterpret_cast<IPv4Range const*>(vb);
+
+            if (*a < b->begin_)
+            {
+                return -1;
+            }
+
+            if (*a > b->end_)
+            {
+                return 1;
+            }
+
+            return 0;
+        }
+    };
+
+    void ensureLoaded();
+    void load();
+    static bool parseLine(char const* line, IPv4Range* range);
+    static bool compareAddressRangesByFirstAddress(IPv4Range const& a, IPv4Range const& b);
+
+    static bool parseLine1(std::string_view line, struct IPv4Range* range);
+    static bool parseLine2(std::string_view line, struct IPv4Range* range);
```

Comment:
iwyu: need `#include <string_view>` at top of header

Reply:
Adding